File: packages/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationManager.java
package com.swmansion.reanimated.keyboard;

import android.app.Dialog;
import com.facebook.react.bridge.ReactApplicationContext;
import java.lang.ref.WeakReference;
import java.util.concurrent.ConcurrentHashMap;

@FunctionalInterface
interface NotifyAboutKeyboardChangeFunction {
  void call();
}

public class KeyboardAnimationManager {
  private int mNextListenerId = 0;
  private final ConcurrentHashMap<Integer, KeyboardWorkletWrapper> mListeners =
      new ConcurrentHashMap<>();
  private final Keyboard mKeyboard = new Keyboard();
  private final ModalManager mModalManager;

  private final WindowInsetsManager mWindowInsetsManager;

  public KeyboardAnimationManager(WeakReference<ReactApplicationContext> reactContext) {
    mWindowInsetsManager =
        new WindowInsetsManager(reactContext, mKeyboard, this::notifyAboutKeyboardChange);
    mModalManager = new ModalManager(reactContext, mKeyboard, this::notifyAboutKeyboardChange);
  }

  public int subscribeForKeyboardUpdates(
      KeyboardWorkletWrapper callback, boolean isStatusBarTranslucent) {
    int listenerId = mNextListenerId++;
    if (mListeners.isEmpty()) {
      mWindowInsetsManager.startObservingChanges(
          this::notifyAboutKeyboardChange, isStatusBarTranslucent);
      mModalManager.startObservingChanges(isStatusBarTranslucent);
    }
    mListeners.put(listenerId, callback);
    return listenerId;
  }

  public void unsubscribeFromKeyboardUpdates(int listenerId) {
    mListeners.remove(listenerId);
    if (mListeners.isEmpty()) {
      mWindowInsetsManager.stopObservingChanges();
      mModalManager.stopObservingChanges();
    }
  }

  public void notifyAboutKeyboardChange() {
    for (KeyboardWorkletWrapper listener : mListeners.values()) {
      listener.invoke(mKeyboard.getState().asInt(), mKeyboard.getHeight());
    }
  }

  public void registerNewDialog(Dialog dialog) {
    mModalManager.registerNewDialog(dialog, !mListeners.isEmpty());
  }
}

File: packages/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationManager.kt
package com.swmansion.reanimated.keyboard

import com.facebook.react.bridge.ReactApplicationContext
import java.lang.ref.WeakReference
import java.util.concurrent.ConcurrentHashMap

fun interface NotifyAboutKeyboardChangeFunction {
    fun call()
}

class KeyboardAnimationManager(
    reactContext: WeakReference<ReactApplicationContext>,
) {
    private var mNextListenerId = 0
    private val mListeners = ConcurrentHashMap<Int, KeyboardWorkletWrapper>()
    private val mKeyboard = Keyboard()
    private val mWindowsInsetsManager =
        WindowsInsetsManager(reactContext, mKeyboard, ::notifyAboutKeyboardChange)

    fun subscribeForKeyboardUpdates(
        callback: KeyboardWorkletWrapper,
        isStatusBarTranslucent: Boolean,
        isNavigationBarTranslucent: Boolean,
    ): Int {
        val listenerId = mNextListenerId++
        if (mListeners.isEmpty()) {
            val keyboardAnimationCallback =
                KeyboardAnimationCallback(mKeyboard, ::notifyAboutKeyboardChange, isNavigationBarTranslucent)
            mWindowsInsetsManager.startObservingChanges(keyboardAnimationCallback, isStatusBarTranslucent, isNavigationBarTranslucent)
        }
        mListeners[listenerId] = callback
        return listenerId
    }

    fun unsubscribeFromKeyboardUpdates(listenerId: Int) {
        mListeners.remove(listenerId)
        if (mListeners.isEmpty()) {
            mWindowsInsetsManager.stopObservingChanges()
        }
    }

    fun notifyAboutKeyboardChange() {
        for (listener in mListeners.values) {
            listener.invoke(mKeyboard.getState().asInt(), mKeyboard.getHeight())
        }
    }
}
